[RDY] Pinning

ramirezmike opened the pull request:
#255 

TODO
- [x] pin and unpin apps keybinding (toggle_pin)
- [x] show/hide pinned apps keybinding (toggle_view_pinned)
- [x] automatically show everything when leaving work mode/automatically re-pin when re-entering workmode
- [x] quitting nog "unpins" and restores pinned windows
- [x] support saving pin state in the workspaces.grid file
- [x] add documentation
- [x] add configuration of how pinned windows look (should allow hiding title bar?)
- [x] "pin-to-workspace" (pin floating windows to a specific workspace)
- [x] keybinding to hide/show windows that are pinned to specific workspaces
- [x] add a rule setting for global_pinned/workspace_pinned like there is for manage
- [x] change API to match latest conversation notes
- [x] streamline API
- [x] refactor pinning code into separate file
- [x] rename find_mut_grid_containing_window to find_grid_containing_window_mut
- [x] update window state when modifying title bar/border
- [x] remove show window logic when modifying title bar/border

TimUntersberger commented:
> add configuration of how pinned windows look (should allow hiding title bar?)

Can you still move the window by using the mouse when hiding the title bar?

ramirezmike commented:
>Can you still move the window by using the mouse when hiding the title bar?

No, you'd have to unpin and move it and then repin. I suspect an approach we could look into is something like... making it so if you have the alt key held down and you click within a pinned window, it acts as if you're moving the window, but I'm not sure how easy that is. I think it's possible though. A "round one" implementation would just be to not have that piece in yet.

That said, I do like the configuration ideas. Would be more flexible.

ramirezmike commented:
So, I pretty much finished this and then re-read the comments and realized I forgot about this whole thing 🤦‍♂️

```
local win = nog.win_get_focused()
nog.win_hide_title_bar(win)
nog.win_toggle_pin(win)
```

I think I'm gonna go with something like the above. I don't believe we have a "win_get_focused" yet, right? 
I see...
**get_current_win** returns the focused window on the current grid
**get_focused_win_of_display** returns the focused window of the grid on a specific display

but win_get_focused would be like.. getting whatever window that the operating system currently has focused, right?

ramirezmike commented:
I started implementing what I mentioned above but I'm not sure how user friendly it is

```
nog.nbind("alt+shift+o", function ()
    local id = nog.get_focused_win()
    nog.win_pin_to_ws(id)
    nog.win_hide_title_bar(id)
end)
```

I got this to work but it would need more functions that gets the state of the window. "pin_to_ws" behaves like a toggle (like toggle_floating) but hide_title_bar and show_title_bar are distinct functions.. 

You'd end up with something like

```
nog.nbind("alt+shift+o", function ()
    local id = nog.get_focused_win()
    nog.win_pin_to_ws(id)
    if nog.win_has_title_bar(id) then
          nog.win_hide_title_bar(id)
    else
          nog.win_show_title_bar(id)
end)
```

which is a little verbose.. What do you think? @TimUntersberger

TimUntersberger commented:
> I got this to work but it would need more functions that gets the state of the window

I created a function for workspace that is called `ws_get_info` (I think?) that returns a table of information about the workspace. We could do the same with windows.

In the future I'd like to play a bit more with metatables so that we maybe could create a Window and Workspace proxy similar to the config.

What it could look like
```lua
local win = nog.get_focused_win()
print(win.title, win.width, win.height, ...)

win.add_titlebar()
win.remove_titlebar()
win.has_titlebar()
...

local ws = nog.get_focused_ws()
ws.add(win)
...
```

> which is a little verbose.. What do you think? @TimUntersberger

The verbosity is not a problem for now, because we can always extract functionality into the `runtime.lua` file for this.
So as long as it is possible on the lua side it should be fine.

ramirezmike commented:
I added a couple items to the public API but I also saw you're making some changes so let me know if I should remove some of these. Also, I added the ability to pin to specific workspaces which I call a "workspace pin" but I don't know what to call non-workspace pin... Would "global pin" be confusing?

```
nog.toggle_view_pinned() -- toggles visibility of globally pinned windows
nog.has_pinned() -- returns whether there are any globally pinned windows
nog.get_focused_win() -- returns the window ID of the operating-system focused window 
nog.win_hide_title_bar(win_id) -- hides the title bar of the given window
nog.win_show_title_bar(win_id) -- shows the title bar of the given window
nog.win_hide_border(win_id) -- hides the border of the given window
nog.win_show_border(win_id) -- shows the border of the given window
nog.win_toggle_pin(win_id) -- toggles global  pin of the given window
nog.win_toggle_ws_pin(win_id) -- toggles pin of the given window on the current workspace
nog.win_is_pinned(win_id) -- returns whether the given window is pinned
nog.ws_toggle_view_pinned() -- toggles the visibility of pinned windows on the active workspace
nog.ws_has_pinned(ws_id) -- returns whether the given workspace has pinned windows
```

Also, I'm thinking of pulling the pinning code into a separate file. Is "pinned" or "pinning" going to be confusing with Rust's concept of Pin? 

@TimUntersberger

TimUntersberger commented:
> I added a couple items to the public API but I also saw you're making some changes so let me know if I should remove some of these

My API pr is currently just experimental.

> Also, I added the ability to pin to specific workspaces which I call a "workspace pin" but I don't know what to call non-workspace pin... Would "global pin" be confusing?

Global should be fine I think

> Also, I'm thinking of pulling the pinning code into a separate file. Is "pinned" or "pinning" going to be confusing with Rust's concept of Pin?

I don't think so, since pinning has a special meaning in our domain and we don't use rust pins.

ramirezmike commented on book/src/configuration/rules.md:
```diff
@@ -17,7 +17,7 @@ A rule can contain the following settings
 |---------------------------|---------|-------------------------------------------------------------------------------|
 | has_custom_titlebar       | Boolean | Changes how we align the window (applications like [vscode](https://code.visualstudio.com/) should have this enabled) |
-| ignore                    | Boolean | Tells nog to not manage this window |
+| action                    | String  | Tells nog how to behave when a new window is created. The value is expected to be one of the following: <ul> <li>**Ignore** - nog will ignore the window</li> <li>**Manage** - nog will manage the window and add it to the currently active workspace</li> <li>**Pin** - nog will globally pin the window (see [pinning mode](/getting-started/pinning_mode.html))</li ...
```
I updated this and also added "Pin" as an option to make a program automatically globally pin when launched.

ramirezmike replied:
I suspect we'll conflict on this line

ramirezmike commented on twm/src/lua/mod.rs:
```diff
@@ -625,12 +625,25 @@ fn setup_nog_global(state_arc: Arc<Mutex<AppState>>, rt: &LuaRuntime) {
-                    .get_active_grids();
-
-                Ok(grids.iter().map(|g| g.id).collect::<Vec<_>>())
+                let display = state_g.get_display_by_id(DisplayId(display_id))
```
this is for the workspace numbers at the top. It's a little verbose but it's getting the IDs that belong to the current display and then filtering it to just the workspaces that either have a populated grid or a pinned window

ramirezmike commented on twm/src/pinned.rs:
```diff
@@ -0,0 +1,257 @@
+use std::collections::HashMap;
```
There's a bit going on in this file but basically it's a separate struct that holds a list of "containers" for pinned windows (one for each workspace and a global one) and handles the hiding/showing of those windows.

ramirezmike commented:
I think this PR is finally in a state that it can be reviewed. I'd recommend filtering on just .md/.rs files. 

Let me know what you think @TimUntersberger , I'm open to suggestions.

I added usage notes in the issue #255

TimUntersberger commented on twm/src/main.rs:
```diff
@@ -720,6 +749,58 @@ impl AppState {
+
+        window.add_title_bar()?;
+
+        if window.is_hidden() {
```
I don't think we should show the window if it is hidden. We might want to add the titlebar without showing it.

ramirezmike replied:
That's a good point, I'll take this out and verify that it behaves correctly.

TimUntersberger commented on twm/src/main.rs:
```diff
@@ -720,6 +749,58 @@ impl AppState {
 
+    pub fn show_title_bar(&mut self, window_id: i32) -> SystemResult {
+        let window_id: WindowId = window_id.into();
+        let mut window: NativeWindow = NativeWindow::from(window_id);
```
Also is it alright that you don't modify the actual window in state? Won't the state get desynced?

ramirezmike replied:
yeah, that's a good point. That shouldn't be too crazy to grab and update.

TimUntersberger commented on twm/src/main.rs:
```diff
@@ -864,18 +964,31 @@ impl AppState {
-    /// Returns the grid containing the window and its corresponding tile
-    pub fn find_grid_containing_window(&mut self, id: WindowId) -> Option<&mut TileGrid> {
+    /// Returns mutable grid containing the window and its corresponding tile
+    pub fn find_mut_grid_containing_window(&mut self, id: WindowId) -> Option<&mut TileGrid> {
```
I think `find_grid_containing_window_mut` is a more idiomatic rust name.

ramirezmike replied:
I actually wrote that out at first and then thought "does that make it sound like the window is mutable and not the grid?" 

The Option<&mut TileGrid> should make it clear though.. I can change this

TimUntersberger commented:
LGTM! Thank you for your hard work!

ramirezmike commented on twm/src/event_handler/winevent.rs:
```diff
@@ -33,9 +35,13 @@ pub fn handle(state: &mut AppState, ev: WinEvent) -> SystemResult {
 
+    // checking global pinned
+    if let Some(window) = state.pinned.get(&ev.window.id.into(), None) {
+        title = Some(window.title.clone());
```
just added a small change here. I was checking the stored pinned and tiled but was basing it off of the grids. This just also checks the global pinned store too

ramirezmike commented on twm/src/pinned.rs:
```diff
@@ -266,9 +266,13 @@ impl Pinned {
 
     pub fn remove(&mut self, window_id: &i32) -> Option<NativeWindow> {
-        for container in self.containers.iter_mut() {
+        for (container_id, container) in self.containers.iter_mut().enumerate() {
```
and this is to update the Store file when a pinned window is removed
